src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -332,86 +340,89 @@ private struct TagEnumerationState : IActivityEnumerator<KeyValuePair<string, ob
 
             public long Port { get; set; }
 
-            public bool ForEach(KeyValuePair<string, object> activityTag)
+            public void EnumerateTags(Activity activity, int maxTags)
             {
-                if (activityTag.Value == null)
+                foreach (ref readonly var tag in activity.EnumerateTagObjects())
                 {
-                    return true;
-                }
-
-                var key = activityTag.Key;
-
-                switch (key)
-                {
-                    case SpanAttributeConstants.StatusCodeKey:
-                        this.StatusCode = activityTag.Value as string;
-                        return true;
-                    case SpanAttributeConstants.StatusDescriptionKey:
-                        this.StatusDescription = activityTag.Value as string;
-                        return true;
-                }
-
-                if (!this.Created)
-                {
-                    this.Tags = PooledList<OtlpCommon.KeyValue>.Create();
-                    this.Created = true;
-                }
+                    if (tag.Value == null)
+                    {
+                        continue;
+                    }
 
-                if (OtlpKeyValueTransformer.Instance.TryTransformTag(activityTag, out var attribute, SdkConfiguration.Instance.AttributeValueLengthLimit))
-                {
-                    PooledList<OtlpCommon.KeyValue>.Add(ref this.Tags, attribute);
+                    var key = tag.Key;
 
-                    if (attribute.Value.ValueCase == OtlpCommon.AnyValue.ValueOneofCase.StringValue)
+                    switch (key)
                     {
-                        PeerServiceResolver.InspectTag(ref this, key, attribute.Value.StringValue);
+                        case SpanAttributeConstants.StatusCodeKey:
+                            this.StatusCode = tag.Value as string;
+                            continue;
+                        case SpanAttributeConstants.StatusDescriptionKey:
+                            this.StatusDescription = tag.Value as string;
+                            continue;
                     }
-                    else if (attribute.Value.ValueCase == OtlpCommon.AnyValue.ValueOneofCase.IntValue)
+
+                    if (OtlpKeyValueTransformer.Instance.TryTransformTag(tag, out var attribute, SdkConfiguration.Instance.AttributeValueLengthLimit))
                     {
-                        PeerServiceResolver.InspectTag(ref this, key, attribute.Value.IntValue);
+                        if (this.Span.Attributes.Count < maxTags)
```

Comment:
@alanwest I think there was a subtle bug in how the limits were implemented before. Previously if we hit the limit, we would break the enumeration. That could lead to the inspection logic getting bypassed and thinks like peer service resolution breaking. I changed the logic so it always enumerates fully, but doesn't add to the lists once the max is reached. That also allowed for drop counting.

Reply:
Yes, I think you're right I introduced a bug which had the potential of skipping the peer service resolution... however, now that I'm thinking more deeply about this peer service resolution, I'm wondering if we need to reevaluate what we're doing here. Or maybe I just need a refresher 😄.

I understand the Zipkin and/or Jaeger specs has some special handling of `peer.service`, though I'm not sure the OTLP exporter has quite the same requirement.

There's the general guidelines [here](https://github.com/open-telemetry/opentelemetry-specification/blob/d1c24937aad93b8d91c3b677976ccb8f432df5b0/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes), but it seems to me that setting `peer.service` would be best done by instrumentation and not by the exporter.

The PeerServiceResolver also suggests to me this is more of an instrumentation concern. Based on the guidelines above, we're missing the mapping of `rpc.service` -> `peer.service`. I think I'd prefer to fix that in gRPC instrumentation rather than updating the PeerServiceResolver.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/28473bc46565b95f6d7d885fb9e669707d138216/src/OpenTelemetry/Internal/PeerServiceResolver.cs#L26-L33

This is all beyond the scope of this PR, but what're your thoughts?

Reply:
I'm not opposed to moving it into instrumentation. Probably means some will not do it, or do it incorrectly, but I don't know if that is necessarily a problem.

Why is OtlpExporter doing it? Really just so it can map cleanly to Jaeger/Zipkin format downstream if needed. Jaeger treats spans without `peer.service` as internal if all services don't report into the same Jaeger instance. Zipkin uses `remoteEndpoint` for that (IIRC) which _should_ be mapped from `peer.service` when converting between the two formats.

Reply:
The Zipkin spec describes the conversion between formats https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/zipkin.md#remote-endpoint

.. but to me it seems this conversion could be performed by a downstream component and not the concern of an upstream OTLP exporter anticipating a Zipkin component downstream.

Anyways, something I might take a stab at following up on. I probably need to spend more time understanding the spec.

Reply:
> Anyways, something I might take a stab at following up on. I probably need to spend more time understanding the spec.

👍